File: android_app/app/src/main/java/com/oceansentinels/app/data/repository/IncidentRepositoryImpl.kt
package com.oceansentinels.app.data.repository

import com.oceansentinels.app.data.local.database.dao.IncidentDao
import com.oceansentinels.app.data.local.database.entity.IncidentEntity
import com.oceansentinels.app.data.remote.api.OceanSentinelsApi
import com.oceansentinels.app.data.remote.dto.CreateIncidentRequestDto
import com.oceansentinels.app.domain.model.*
import com.oceansentinels.app.domain.repository.IncidentRepository
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import timber.log.Timber
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Implementation of IncidentRepository
 */
@Singleton
class IncidentRepositoryImpl @Inject constructor(
    private val api: OceanSentinelsApi,
    private val incidentDao: IncidentDao
) : IncidentRepository {

    override suspend fun getIncidents(filters: IncidentFilters): Result<IncidentListResult> {
        return try {
            Timber.d("Fetching incidents with filters: $filters")
            
            val response = api.getIncidents(
                page = filters.page,
                size = filters.size,
                status = filters.status?.value,
                hazardType = filters.hazardType?.value,
                urgency = filters.urgency?.value,
                search = filters.searchQuery
            )
            
            if (response.isSuccessful && response.body() != null) {
                val result = response.body()!!.toDomain()
                
                // Cache incidents
                val entities = result.incidents.map { IncidentEntity.fromDomain(it) }
                incidentDao.insertAll(entities)
                
                Timber.d("Fetched ${result.incidents.size} incidents")
                Result.success(result)
            } else {
                val errorBody = response.errorBody()?.string() ?: "Unknown error"
                Timber.e("Failed to fetch incidents: $errorBody")
                Result.failure(Exception("Failed to fetch incidents: $errorBody"))
            }
        } catch (e: Exception) {
            Timber.e(e, "Error fetching incidents")
            // Try to return cached data
            val cached = getCachedIncidents()
            if (cached.isNotEmpty()) {
                Result.success(
                    IncidentListResult(
                        incidents = cached,
                        total = cached.size,
                        page = 1,
                        size = cached.size,
                        hasNext = false,
                        hasPrev = false
                    )
                )
            } else {
                Result.failure(e)
            }
        }
    }

    override fun getIncidentsFlow(filters: IncidentFilters): Flow<Result<IncidentListResult>> = flow {
        emit(getIncidents(filters))
    }

    override suspend fun getIncident(id: Int): Result<Incident> {
        return try {
            val response = api.getIncident(id)
            
            if (response.isSuccessful && response.body() != null) {
                val incident = response.body()!!.toDomain()
                incidentDao.insert(IncidentEntity.fromDomain(incident))
                Result.success(incident)
            } else {
                // Try cache
                val cached = incidentDao.getIncidentById(id)
                if (cached != null) {
                    Result.success(cached.toDomain())
                } else {
                    Result.failure(Exception("Incident not found"))
                }
            }
        } catch (e: Exception) {
            val cached = incidentDao.getIncidentById(id)
            if (cached != null) {
                Result.success(cached.toDomain())
            } else {
                Result.failure(e)
            }
        }
    }

    override suspend fun getIncidentByReference(referenceId: String): Result<Incident> {
        return try {
            val cached = incidentDao.getIncidentByReferenceId(referenceId)
            if (cached != null) {
                Result.success(cached.toDomain())
            } else {
                Result.failure(Exception("Incident not found"))
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    override suspend fun getMyReports(filters: IncidentFilters): Result<IncidentListResult> {
        // The API filters by user automatically based on auth token
        return getIncidents(filters)
    }

    override suspend fun createIncident(request: CreateIncidentRequest): Result<Incident> {
        return try {
            Timber.d("Creating new incident: ${request.hazardType}")
            
            val dto = CreateIncidentRequestDto.fromDomain(request)
            val response = api.createIncident(dto)
            
            if (response.isSuccessful && response.body() != null) {
                val incident = response.body()!!.toDomain()
                incidentDao.insert(IncidentEntity.fromDomain(incident))
                Timber.d("Incident created successfully: ${incident.referenceId}")
                Result.success(incident)
            } else {
                val errorBody = response.errorBody()?.string() ?: "Unknown error"
                Timber.e("Failed to create incident: $errorBody")
                Result.failure(Exception("Failed to create incident: $errorBody"))
            }
        } catch (e: Exception) {
            Timber.e(e, "Error creating incident")
            Result.failure(e)
        }
    }

    override suspend fun verifyIncident(id: Int): Result<Incident> {
        return try {
            Timber.d("Verifying incident: $id")
            
            val response = api.verifyIncident(id)
            
            if (response.isSuccessful) {
                Timber.d("Incident verified successfully: ${response.body()?.message}")
                // Re-fetch the updated incident from server
                return getIncident(id)
            } else {
                val errorBody = response.errorBody()?.string() ?: "Unknown error"
                Result.failure(Exception("Failed to verify incident: $errorBody"))
            }
        } catch (e: Exception) {
            Timber.e(e, "Error verifying incident")
            Result.failure(e)
        }
    }

    override suspend fun deployResponse(id: Int): Result<Incident> {
        return try {
            Timber.d("Deploying response to incident: $id")
            
            val response = api.deployResponse(id)
            
            if (response.isSuccessful) {
                Timber.d("Response deployed successfully: ${response.body()?.message}")
                // Re-fetch the updated incident from server
                return getIncident(id)
            } else {
                val errorBody = response.errorBody()?.string() ?: "Unknown error"
                Result.failure(Exception("Failed to deploy response: $errorBody"))
            }
        } catch (e: Exception) {
            Timber.e(e, "Error deploying response")
            Result.failure(e)
        }
    }

    override suspend fun resolveIncident(id: Int): Result<Incident> {
        return try {
            Timber.d("Resolving incident: $id")
            
            val response = api.resolveIncident(id)
            
            if (response.isSuccessful) {
                Timber.d("Incident resolved successfully: ${response.body()?.message}")
                // Re-fetch the updated incident from server
                return getIncident(id)
            } else {
                val errorBody = response.errorBody()?.string() ?: "Unknown error"
                Result.failure(Exception("Failed to resolve incident: $errorBody"))
            }
        } catch (e: Exception) {
            Timber.e(e, "Error resolving incident")
            Result.failure(e)
        }
    }

    override suspend fun getCachedIncidents(): List<Incident> {
        return incidentDao.getAllIncidentsList().map { it.toDomain() }
    }

    override suspend fun syncIncidents(): Result<Unit> {
        return try {
            val pendingIncidents = incidentDao.getPendingSyncIncidents()
            
            pendingIncidents.forEach { entity ->
                // In a real app, this would sync pending offline changes
                Timber.d("Syncing incident: ${entity.referenceId}")
            }
            
            // Refresh from server
            getIncidents(IncidentFilters())
            Result.success(Unit)
        } catch (e: Exception) {
            Timber.e(e, "Error syncing incidents")
            Result.failure(e)
        }
    }
    
    override suspend fun assignIncident(incidentId: Int, rescueTeamUserId: Int): Result<Unit> {
        return try {
            Timber.d("Assigning incident $incidentId to user $rescueTeamUserId")
            val response = api.assignIncident(incidentId, mapOf("assigned_to_id" to rescueTeamUserId))
            if (response.isSuccessful) {
                Timber.d("Incident assigned successfully")
                Result.success(Unit)
            } else {
                val errorBody = response.errorBody()?.string() ?: "Unknown error"
                Result.failure(Exception("Failed to assign incident: $errorBody"))
            }
        } catch (e: Exception) {
            Timber.e(e, "Error assigning incident")
            Result.failure(e)
        }
    }
    
    override suspend fun getAssignedIncidents(page: Int, size: Int): Result<IncidentListResult> {
        return try {
            Timber.d("Fetching assigned incidents page=$page")
            val response = api.getMyAssignedIncidents(page, size)
            if (response.isSuccessful && response.body() != null) {
                val result = response.body()!!.toDomain()
                Timber.d("Fetched ${result.incidents.size} assigned incidents")
                Result.success(result)
            } else {
                val errorBody = response.errorBody()?.string() ?: "Unknown error"
                Result.failure(Exception("Failed to fetch assigned incidents: $errorBody"))
            }
        } catch (e: Exception) {
            Timber.e(e, "Error fetching assigned incidents")
            Result.failure(e)
        }
    }
}

File: IOS/OceanSentinels/Data/Repository/IncidentRepositoryImpl.swift
import Foundation
import os

/// Implementation of IncidentRepository.
/// Replaces Kotlin IncidentRepositoryImpl with URLSession + SwiftData caching.
final class IncidentRepositoryImpl: IncidentRepository {
    
    private let api: OceanSentinelsAPI
    private let database: DatabaseManager
    private let logger = Logger(subsystem: "com.oceansentinels.app", category: "IncidentRepo")
    
    init(api: OceanSentinelsAPI, database: DatabaseManager) {
        self.api = api
        self.database = database
    }
    
    func getIncidents(filters: IncidentFilters) async throws -> IncidentListResult {
        logger.debug("Fetching incidents with filters")
        
        do {
            let response = try await api.getIncidents(
                page: filters.page,
                size: filters.size,
                status: filters.status?.value,
                hazardType: filters.hazardType?.value,
                urgency: filters.urgency?.value,
                search: filters.searchQuery
            )
            
            let result = response.toDomain()
            
            // Cache incidents
            let entities = result.incidents.map { IncidentEntity.fromDomain($0) }
            try await database.insertIncidents(entities)
            
            logger.debug("Fetched \(result.incidents.count) incidents")
            return result
        } catch {
            logger.error("Error fetching incidents: \(error)")
            // Try to return cached data
            let cached = try await getCachedIncidents()
            if !cached.isEmpty {
                return IncidentListResult(
                    incidents: cached,
                    total: cached.count,
                    page: 1,
                    size: cached.count,
                    hasNext: false,
                    hasPrev: false
                )
            }
            throw error
        }
    }
    
    func getIncident(id: Int) async throws -> Incident {
        do {
            let dto = try await api.getIncident(id: id)
            let incident = dto.toDomain()
            try await database.insertIncident(IncidentEntity.fromDomain(incident))
            return incident
        } catch {
            // Try cache
            if let cached = try await database.fetchIncident(byId: id) {
                return cached.toDomain()
            }
            throw error
        }
    }
    
    func getIncidentByReference(referenceId: String) async throws -> Incident {
        if let cached = try await database.fetchIncident(byReferenceId: referenceId) {
            return cached.toDomain()
        }
        throw NSError(domain: "IncidentRepo", code: 404, userInfo: [NSLocalizedDescriptionKey: "Incident not found"])
    }
    
    func getMyReports(filters: IncidentFilters) async throws -> IncidentListResult {
        // API filters by user automatically based on auth token
        try await getIncidents(filters: filters)
    }
    
    func createIncident(request: CreateIncidentRequest) async throws -> Incident {
        logger.debug("Creating new incident: \(request.hazardType.rawValue)")
        
        let dto = CreateIncidentRequestDTO.fromDomain(request)
        let responseDTO = try await api.createIncident(request: dto)
        let incident = responseDTO.toDomain()
        
        try await database.insertIncident(IncidentEntity.fromDomain(incident))
        
        logger.debug("Incident created successfully: \(incident.referenceId)")
        return incident
    }
    
    func verifyIncident(id: Int) async throws -> Incident {
        logger.debug("Verifying incident: \(id)")
        _ = try await api.verifyIncident(id: id)
        return try await getIncident(id: id)
    }
    
    func deployResponse(id: Int) async throws -> Incident {
        logger.debug("Deploying response to incident: \(id)")
        _ = try await api.deployResponse(id: id)
        return try await getIncident(id: id)
    }
    
    func resolveIncident(id: Int) async throws -> Incident {
        logger.debug("Resolving incident: \(id)")
        _ = try await api.resolveIncident(id: id)
        return try await getIncident(id: id)
    }
    
    func assignIncident(incidentId: Int, rescueTeamUserId: Int) async throws {
        logger.debug("Assigning incident \(incidentId) to user \(rescueTeamUserId)")
        _ = try await api.assignIncident(id: incidentId, userId: rescueTeamUserId)
    }
    
    func getAssignedIncidents(page: Int, size: Int) async throws -> IncidentListResult {
        logger.debug("Fetching assigned incidents page=\(page)")
        let response = try await api.getMyAssignedIncidents(page: page, size: size)
        return response.toDomain()
    }
    
    func getCachedIncidents() async throws -> [Incident] {
        try await database.fetchAllIncidents().map { $0.toDomain() }
    }
    
    func syncIncidents() async throws {
        let pendingIncidents = try await database.fetchPendingSyncIncidents()
        for entity in pendingIncidents {
            logger.debug("Syncing incident: \(entity.referenceId)")
        }
        // Refresh from server
        _ = try await getIncidents(filters: IncidentFilters())
    }
    
    func uploadPhoto(imageData: Data, fileName: String) async throws -> String {
        let response = try await api.uploadIncidentPhoto(imageData: imageData, fileName: fileName)
        return response.url
    }
}
